src/main/java/com/team766/controllers/ExtendableCanSparkMaxSmartMotionPIDController.java

```diff
@@ -0,0 +1,291 @@
+	 */
+	public void setOutputRange(final double min, final double max) {
+		pid.setOutputRange(min, max);
+	}
+
+	/*
+	 * This is the method to set the minimum and maximum locations of the mechanism.
+	 * Please check your subclasses documentation for what units this is supposed to be in (usually whichever one is used as an input for the setSetpoint method)
+	 * @author Max Spier - 9/9/2023
+	 */
+	public void setMinMaxLocation(final double min, final double max) {
+		maxPos = max;
+		minPos = min;
+	}
+
+	/*
+	* This method is used to set the maximum velocity of the mechanism for use with PID
+	* @param max the value of the maximum amount of velocity for the motorcontroller
+	* The method also sets the minimum velocity to zero, as a failsafe.
+	* @author Max Spier - 9/9/2023
+	*/
+	public void setMaxVelocity(final double max) {
+		pid.setSmartMotionMaxVelocity(max, 0);
+		pid.setSmartMotionMinOutputVelocity(0, 0);
+	}
+
+	/*
+	 * This is the method to set the maximum acelleration of the robot for the PIDController
+	 * @param max the value for the maximum acelleration
+	 * @author Max Spier - 9/9/2023
+	 */
+	public void setMaxAccel(final double max) {
+		pid.setSmartMotionMaxAccel(max, 0);
+	}
+
+	/*
+	 * This is the method where you update the setpoint position from the subclass
+	 * @param x the setpoint
+	 * @author Max Spier - 9/9/2023
+	 */
+	public void setNewSetpoint(double x){
+		x = MathUtil.clamp(x, minPos, maxPos);
+		setPointPosition = x;
+	}
+
+	/*
+	 * Failsafe for the PIDController to stop it.
+	 * @author Max Spier - 9/9/2023
+	 */
+	public void stop() {
+		theState = PIDState.OFF;
+	}
+
+	/*
+	 * This method does the antigrav
+	 * @author Max Spier - 9/9/2023
+	 */
+	private void antigrav(){
+		mc.set(curAntigrav);
+	}
+
+	/*
+	 * This method sets the antigrav power amount.
+	 * It gets the value for this from the subclass
+	 * @param passedValueFromSubclass the power to set the motorcontroller for antigrav
+	 * @author Max Spier - 9/9/2023
+	 */
+	protected void updateAntigrav(double passedValueFromSubclass){
+		curAntigrav = passedValueFromSubclass;
+	}
+
+	/*
+	 * This is the run loop that actually runs the PID Controller
+	 * You need to call this as frequently as possible when running the mechanism
+	 * @author Max Spier - 9/9/2023
+	 */
+	public void run() {
+		switch (theState) {
+			case OFF:
+				mc.set(0);
+				break;
+			case ANTIGRAV:
+				if (setPointPosition <= (deadzone + mc.getSensorPosition()) && setPointPosition >= (mc.getSensorPosition() - deadzone)) {
+					antigrav();
+				} else {
+					theState = PIDState.PID;
+				}
+			case PID:
+				if (setPointPosition <= (deadzone + mc.getSensorPosition()) && setPointPosition >= (mc.getSensorPosition() - deadzone)) {
+					comboOfTimesInsideDeadzone++;
+				} else {
+					comboOfTimesInsideDeadzone = 0;
+					pid.setReference(setPointPosition, ControlType.kSmartMotion);
+				}
+
+				if (comboOfTimesInsideDeadzone >= 6) {
+					theState = PIDState.ANTIGRAV;
```

Comment:
we need a flag to disable antigrav if its not needed

also, I realized Brake mode is missing

Reply:
finished brake

Reply:
finished all

Reply:
I think you did not push your changes

Reply:
no i did

Reply:
(https://github.com/Team766/MaroonFramework/pull/27/files/cb3c9486136cf3cbbf7be467f6b16e9fd9044a3c..4724223a8237d85820d3079245a7026d1395b1d6)

Reply:
how does it look?